code/game/objects/structures/crematorium.dm

```diff
@@ -0,0 +1,262 @@
+/obj/structure/crematorium
+	name = "crematorium"
+	desc = "A human incinerator. Works well on barbeque nights."
+	icon = 'icons/obj/structures/crematorium.dmi'
+	icon_state = "crematorium_closed"
+	density = TRUE
+	anchored = TRUE
+
+	var/cremating = FALSE
+	var/locked = FALSE
+	var/open = FALSE
+
+	var/obj/structure/crematorium_tray/connected_tray
+
+	var/id_tag
+
+/obj/structure/crematorium/Initialize(ml, _mat, _reinf_mat)
+	. = ..()
+	connected_tray = new /obj/structure/crematorium_tray(src)
+	connected_tray.connected_crematorium = src
+
+/obj/structure/crematorium/Destroy()
+	if(!QDELETED(connected_tray))
+		QDEL_NULL(connected_tray)
+	return ..()
+
+/obj/structure/crematorium/on_update_icon()
+	if(cremating)
+		icon_state = "crematorium_active"
+	else if (open)
+		icon_state = "crematorium_open"
+	else if (contents.len > 1)
+		icon_state = "crematorium_filled"
+	else
+		icon_state = "crematorium_closed"
+
+/obj/structure/crematorium/explosion_act(severity)
+	..()
+	if(!QDELETED(src) && (severity == 1 || (severity == 2 && prob(50)) || (severity == 3 && prob(5))))
+		physically_destroyed()
+
+/obj/structure/crematorium/proc/open()
+	if(cremating || locked || open)
+		return
+
+	if(!connected_tray)
+		return
+
+	playsound(src, 'sound/items/Deconstruct.ogg', 50, 1)
+	var/turf/T = get_step(src, dir)
+	connected_tray.forceMove(T)
+	connected_tray.set_dir(dir)
+	for(var/atom/movable/A in src)
+		if(!ismob(A) && !isobj(A))
```

Comment:
This should be checking `simulated` and/or `anchored`, not relying on the object not being a mob or obj.